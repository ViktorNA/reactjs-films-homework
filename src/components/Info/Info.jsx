import propTypes from 'prop-types';
import React, { Component } from 'react';
import uniqid from 'uniqid';
import styles from './info.scss';

const baseUrl = 'https://image.tmdb.org/t/p/original/';

const mainMovie = require('../../assets/json/505954.json');

export default class Info extends Component {
  getGenres = (genres, duration) => {
    const dur = `${Math.floor(duration / 60)}h ${duration % 60}m`;
    return (
      <div>
        {genres.map(genre => (
          <div className={styles.item} key={uniqid()}>{genre.name}</div>
        ))}
        <div className={styles.item}>|</div>
        <div className={styles.item}>{dur}</div>
      </div>
    );
  }

  getRating = (rating) => {
    const stars = Math.ceil(rating);
    const array = Array.from({ length: stars }, x => x);
    return (
      <div className={styles.starBox}>
        {array.map(() => (
          <div className={styles.star} key={uniqid()} />
        ))}
      </div>
    );
  }

  render() {
    const {
      vote_average: voteAverage,
      title,
      genres,
      runtime,
      overview,
      backdrop_path: backdropPath,
    } = mainMovie;
    return (
      <div style={{ backgroundImage: `url(${baseUrl + backdropPath})` }} className={styles.wrapper}>
        <div className={styles.bottomShadow} />
        <div className={styles.infoBox}>
          <div className={styles.movieName}>{title}</div>
          <div className={styles.genresBox}>{this.getGenres(genres, runtime)}</div>
          <div className={styles.ratingBox}>
            {this.getRating(voteAverage)}
            <div className={styles.rating}>{voteAverage}</div>
          </div>
        </div>
        <div className={styles.descBox}>
          <div className={styles.description}>
            <div className={styles.textWrapper}>{overview}</div>
          </div>
          <button type="button" className={styles.watchNowButton}>Watch Now</button>
          <button type="button" className={styles.viewInfoButton}>View Info</button>
        </div>
      </div>
    );
  }
}

Info.propTypes = {
  movie: propTypes.shape({
    vote_average: propTypes.number.isRequired,
    title: propTypes.string.isRequired,
    genres: propTypes.array.isRequired,
    runtime: propTypes.number.isRequired,
    overview: propTypes.string.isRequired,
  }),
};

Info.defaultProps = {
  movie: {
    title: 'error',
    overview: 'error',
    vote_average: 0,
    genres: ['uncnown genre'],
    runtime: 0,
  },
};
