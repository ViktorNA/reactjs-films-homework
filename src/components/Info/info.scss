.wrapper {
  display: grid;
  grid-template: 100px 1fr 200px/1fr 1fr;
  position: relative;
  width: 100%;
  height: 500px;
  background-size: cover;
  background-position: center;
  margin-bottom: 38px;
}

.image {
  width: inherit;
}

.topShadow {
  position: absolute;
  width: inherit;
  height: 100px;
  background-image: linear-gradient(rgba(0, 0, 0, 0.7), rgba(0, 0, 0, 0));
}

.bottomShadow {
  width: inherit;
  position: absolute;
  bottom: 0;
  height: 200px;
  background-image: linear-gradient(rgba(0, 0, 0, 0), rgba(0, 0, 0, 1));
}

.filmsLabel {
  margin-left: 70px;
  margin-top: 15px;
  z-index: 999;
  color: white;
  font-size: 25px;
  text-align: left;
}

.searchBox {
  z-index: 999;
  text-align: right;
}

.searchInput {
  border: 2px solid white;
  border-radius: 6px;
  color: white;
  font-size: 14px;
  padding-left: 10px;
  outline: 0;
  width: 270px;
  height: 30px;
  background-color: rgba(0, 0, 0, 0);
  margin-right: 60px;
  margin-top: 15px;
  background-image: url(../../assets/images/Search.svg) no-repeat scroll 245px 6px;
  background-size: 20px 20px;
}

.searchInput::placeholder {
  color: white;
}

.infoBox {
  z-index: 999;
  grid-row: 3/4;
  margin-left: 70px;
  margin-top: 70px;
}

.movieName {
  color: #fff;
  font-size: 30px;
  text-align: left;
  font-weight: bold;
}

.genresBox {
  color: #fff;
  font-size: 12px;
  margin-top: 6px;
}

.item {
  display: inline-block;
  margin-right: 10px;
}

.star {
  position: relative;
  display: inline-block;
  width: 0;
  height: 0;
  margin-left: 0.9em;
  margin-right: 0.9em;
  margin-bottom: 1.2em;
  border-right: 0.3em solid transparent;
  border-bottom: 0.7em  solid #0aaee4;
  border-left: 0.3em solid transparent;
  font-size: 10px;

  &::before,
  &::after {
    content: '';
    display: block;
    width: 0;
    height: 0;
    position: absolute;
    top: 0.6em;
    left: -1em;
    border-right: 1em solid transparent;
    border-bottom: 0.7em  solid #0aaee4;
    border-left: 1em solid transparent;
    transform: rotate(-35deg);
  }

  &::after {
    transform: rotate(35deg);
  }
}

.starBox {
  display: inline-block;
}

.rating {
  position: absolute;
  display: inline-block;
  margin-left: 5px;
  color: #fff;
  font-size: 12px;
  text-align: center;
  border-radius: 3px;
  border: 2px solid #0aaee4;
  width: 28px;
  height: 16px;
}

.ratingBox {
  position: relative;
  margin-top: 23px;
}

.descBox {
  grid-row: 3/4;
  z-index: 999;
  height: 300px;
  justify-self: end;
  justify-items: right;
  align-items: end;
  margin-right: 10px;
}

.description {
  background-color: rgba(0, 0, 0, 0.64);
  width: 230px;
  height: 100px;
  padding: 15px;
  margin-bottom: 23px;
  margin-left: 50px;
}

.textWrapper {
  font-size: 13px;
  color: white;
  line-height: 1.6em;
  height: 100px;
  overflow: hidden;
}

.watchNowButton {
  font-size: 14px;
  color: white;
  border: 0;
  border-radius: 3px;
  background-color: #3bb4e0;
  width: 100px;
  height: 30px;
  margin-right: 21px;
}

.viewInfoButton {
  font-size: 14px;
  color: white;
  border-radius: 3px;
  border: 1px solid white;
  width: 100px;
  height: 28px;
  background-color: rgba(255, 255, 255, 0);
  align-content: center;
  margin-right: 70px;
}
