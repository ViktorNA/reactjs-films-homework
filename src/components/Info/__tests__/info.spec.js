import TestRenderer from 'react-test-renderer';
import React from 'react';
import Info from '../Info';

describe('Snapshot test', () => {
  it('works correctly', () => {
    const testRender = TestRenderer.create(<Info />);
    expect(testRender.toJSON()).toMatchSnapshot();
  });
});

describe('channgeState', () => {
  it('works correctly', () => {
    const testRender = TestRenderer.create(<Info />);
    const testInstance = testRender.getInstance();
    const res = testInstance.getRating(5);
    expect(res.props.children.length).toEqual(5);
  });
});
