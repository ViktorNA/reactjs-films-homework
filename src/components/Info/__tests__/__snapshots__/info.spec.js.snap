// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test works correctly 1`] = `
<div
  className="wrapper"
  style={
    Object {
      "backgroundImage": "url(https://image.tmdb.org/t/p/original//59qyrVGLLtrcuFtJGdixzn7H4OI.jpg)",
    }
  }
>
  <div
    className="bottomShadow"
  />
  <div
    className="infoBox"
  >
    <div
      className="movieName"
    >
      T-34
    </div>
    <div
      className="genresBox"
    >
      <div>
        <div
          className="item"
        >
          War
        </div>
        <div
          className="item"
        >
          Drama
        </div>
        <div
          className="item"
        >
          Adventure
        </div>
        <div
          className="item"
        >
          Action
        </div>
        <div
          className="item"
        >
          |
        </div>
        <div
          className="item"
        >
          2h 19m
        </div>
      </div>
    </div>
    <div
      className="ratingBox"
    >
      <div
        className="starBox"
      >
        <div
          className="star"
        />
        <div
          className="star"
        />
        <div
          className="star"
        />
        <div
          className="star"
        />
        <div
          className="star"
        />
      </div>
      <div
        className="rating"
      >
        4.7
      </div>
    </div>
  </div>
  <div
    className="descBox"
  >
    <div
      className="description"
    >
      <div
        className="textWrapper"
      >
        1941 - WWii. the second lieutenant Nikolai ivushkin, commander of a t-34, engages in an unequal battle against the tank ace Klaus Jager in a battle near moscow. His mission is more of a suicide - to destroy a dozen german tanks, all by himself. that said, luck does favour the bold. He wins the battle, barely survives, but loses his tank and lands himself in captivity for three long years... there was little to no chance for ivushkin and Jager to meet again, but the war knows how to throw a curve ball.in the spring of 1944, the Wehrmacht commands Jager to take charge of the ohrdruf  re range and turn it into a training center for elite german armored forces, using the latest t-34 as a running target. this is how Jager and ivushkin cross paths again. Jager o ers ivushkin to become the commander of a legendary tank and pick his crew from fellow camp prisoners. Nothing goes according to plan, though, when ivushkin uses exercises for a daring and carefully planned escape.
      </div>
    </div>
    <button
      className="watchNowButton"
      type="button"
    >
      Watch Now
    </button>
    <button
      className="viewInfoButton"
      type="button"
    >
      View Info
    </button>
  </div>
</div>
`;
